File: Problem4.java
// ## Problem4  (https://leetcode.com/problems/intersection-of-two-linked-lists/)
// Time Complexity : O(M+N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


public class Solution {
    // Time: O(M+N)
    // Space: O(1)
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lenA=0, lenB=0;

        // Find length
        ListNode h1=headA;
        ListNode h2=headB;

        while(h1!=null){
            lenA++;
            h1=h1.next;
        }

        while(h2!=null){
            lenB++;
            h2=h2.next;
        }

        // Take longer of lengths and move it until no difference in length
        h1=headA;
        h2=headB;
        if(lenA>lenB){
            while(lenA!=lenB){
                h1=h1.next;
                lenA--;
            }
        }else{
            while(lenA!=lenB){
                h2=h2.next;
                lenB--;
            }
        }

        // Move both together until common nodes encountered or both becomes null
        while(h1!=h2){
            h1=h1.next;
            h2=h2.next;
        }

        return h1;
    }
}

File: Problem4.js
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

var getIntersectionNode = function (headA, headB) {
    let current = headA;
    let lenA = 0;
    while (current) {
        current = current.next;
        lenA++;
    }
    current = headB;
    let lenB = 0;
    while (current) {
        current = current.next;
        lenB++;
    }

    while (lenA > lenB) {
        headA = headA.next;
        lenA--;
    }
    while (lenB > lenA) {
        headB = headB.next;
        lenB--;
    }
    while (headA != headB) {
        headA = headA.next;
        headB = headB.next;
    }
    return headA;
};